Blizzard.IO.RabbitMQ/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using Blizzard.IO.Core;
+using Blizzard.IO.RabbitMQ.Entities;
+using EasyNetQ;
+using EasyNetQ.Topology;
+using Microsoft.Extensions.Logging;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public class NetqRabbitConsumer<TData> : IConsumer<TData>, IConsumerWithMetadata<TData, RabbitMessageProperties>
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IAbstractTypeDeserializer<TData> _concreteTypeDeserializer;
+        private readonly IQueue _netqQueue;
+        private readonly IConverter<MessageProperties, RabbitMessageProperties> _converter;
+        private readonly ILogger _logger;
+        
+        private IDisposable _consumeHandler;
+        private readonly Func<byte[], MessageProperties, TData> _dataExtractionStrategy;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        protected NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> coverter)
+        {
+            _netqBus = netqBus;
+            _netqQueue = new Queue(sourceQueue.Name, sourceQueue.Exclusive);
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitConsumer<TData>));
+            _converter = coverter ?? new RabbitPropertiesConverter();
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IDeserializer<TData> deserializer, 
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _deserializer = deserializer;
+            _dataExtractionStrategy += ExtractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} data");
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IAbstractTypeDeserializer<TData> concreteTypeDeserializer,
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _concreteTypeDeserializer = concreteTypeDeserializer;
+            _dataExtractionStrategy += ExtractAbstractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} derieved data");
+        }
+
+        public void Start()
+        {
```

Comment:
Add here a check if the user has subscribed to atleast one event and trow exception if he is not subscribed to at least one event. (in order to prevent the bug your team found)

Reply:
f